rasa/nlu/classifiers/diet_classifier.py

```diff
@@ -551,26 +579,60 @@ def _check_input_dimension_consistency(self, model_data: RasaModelData) -> None:
                 )
 
     def _extract_labels_precomputed_features(
-        self, label_examples: List[Message], attribute: Text = INTENT
+        self, label_examples: List[Message], label_attribute: Text = INTENT
     ) -> Tuple[List[FeatureArray], List[FeatureArray]]:
-        """Collects precomputed encodings."""
-        features = defaultdict(list)
+        """Collects and combines the label_attribute related features from all messages.
 
+        For each message, it extracts the features and combines features of the same
+        feature type and level combination into a single sparse or dense matrix.
+
+        Then, for each feature level and feature type, it turns the extracted
+        matrices (from the various messages) into a single `Features` object.
+
+        Finally, the resuling FeatureArrays are distributed into two lists, one for
+        sequence and one for sentence level features. If both feature types are
+        present within such a list, then sparse features come first.
```

Comment:
As I mention below, as a rule of thumb I think the doc string should just explain what kind of output you should expect, and what properties this output is guaranteed to have. Not how it is computed.

Reply:
I added this comment before I then decided to rewrite the function because it wasn't so obvious to me before - though I think that was mostly because I got to get used to functions concatenating features for me 😅 .